pkg/controller/snapshot_controller.go

```diff
@@ -0,0 +1,773 @@
+		if !IsSnapshotBound(snapshot, content) {
+			// snapshot is bound but content is not bound to snapshot correctly
+			if _, err = ctrl.updateSnapshotErrorStatusWithEvent(snapshot, v1.EventTypeWarning, "SnapshotMisbound", "VolumeSnapshotContent is not bound to the VolumeSnapshot correctly"); err != nil {
+				return err
+			}
+			return nil
+		}
+		// Snapshot is correctly bound.
+		if _, err = ctrl.updateSnapshotBoundWithEvent(snapshot, v1.EventTypeNormal, "SnapshotBound", "Snapshot is bound to its VolumeSnapshotContent"); err != nil {
+			return err
+		}
+		return nil
+	}
+}
+
+func (ctrl *csiSnapshotController) syncUnboundSnapshot(snapshot *crdv1.VolumeSnapshot) error {
+	uniqueSnapshotName := snapshotKey(snapshot)
+	glog.V(4).Infof("syncSnapshot %s", uniqueSnapshotName)
+
+	// Snsapshot has errors during its creation. Controller will not try to fix it. Nothing to do.
+	if snapshot.Status.Error != nil {
+		//return nil
+	}
+
+	if snapshot.Spec.SnapshotContentName != "" {
+		contentObj, found, err := ctrl.contentStore.GetByKey(snapshot.Spec.SnapshotContentName)
+		if err != nil {
+			return err
+		}
+		if !found {
+			// snapshot is bound to a non-existing content.
+			ctrl.updateSnapshotErrorStatusWithEvent(snapshot, v1.EventTypeWarning, "SnapshotLost", fmt.Sprintf("Snapshot has lost reference to VolumeSnapshotContent, %v", err))
+			return fmt.Errorf("snapshot %s is bound to a non-existing content %s", uniqueSnapshotName, snapshot.Spec.SnapshotContentName)
+		}
+		content, ok := contentObj.(*crdv1.VolumeSnapshotContent)
+		if !ok {
+			return fmt.Errorf("expected volume snapshot content, got %+v", contentObj)
+		}
+
+		if err := ctrl.bindSnapshotContent(snapshot, content); err != nil {
+			// snapshot is bound but content is not bound to snapshot correctly
+			ctrl.updateSnapshotErrorStatusWithEvent(snapshot, v1.EventTypeWarning, "SnapshotBoundFailed", fmt.Sprintf("Snapshot failed to bind VolumeSnapshotContent, %v", err))
+			return fmt.Errorf("snapshot %s is bound, but VolumeSnapshotContent %s is not bound to the VolumeSnapshot correctly, %v", uniqueSnapshotName, content.Name, err)
+		}
+
+		// snapshot is already bound correctly, check the status and update if it is ready.
+		glog.V(4).Infof("Check and update snapshot %s status", uniqueSnapshotName)
+		if err = ctrl.checkandUpdateSnapshotStatus(snapshot, content); err != nil {
+			return err
+		}
+		return nil
+	} else { // snapshot.Spec.SnapshotContentName == nil
+		if ContentObj := ctrl.getMatchSnapshotContent(snapshot); ContentObj != nil {
+			glog.V(4).Infof("Find VolumeSnapshotContent object %s for snapshot %s", ContentObj.Name, uniqueSnapshotName)
+			newSnapshot, err := ctrl.bindandUpdateVolumeSnapshot(ContentObj, snapshot)
+			if err != nil {
+				return err
+			}
+			glog.V(4).Infof("bindandUpdateVolumeSnapshot %v", newSnapshot)
+			return nil
+		} else if snapshot.Status.Error == nil { // Try to create snapshot if no error status is set
```

Comment:
This looks suspicious. So we never retry creating a snapshot?

And it looks racy, if a create operation is running in parallel, above we check that there is no error. Just after that the creation operation fails and set the error and finishes. This function resumes and enqueues new creation operation.

IMO, `snapshot.Status.Error` should be checked also in the operation. If it is really necessary.

Reply:
The current design is not retry to create a snapshot. Later we will add a retry policy to allow user specify how long to retry. I can add the check error inside the operation too.